src/mono/netcore/System.Private.CoreLib/src/System/Environment.iOS.cs

```diff
@@ -0,0 +1,98 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using System.Threading;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using NSSearchPathDirectory = Interop.Sys.NSSearchPathDirectory;
+
+namespace System
+{
+    public static partial class Environment
+    {
+        private static Dictionary<SpecialFolder, string> s_specialFolders;
+
+        private static string GetFolderPathCore(SpecialFolder folder, SpecialFolderOption option)
+        {
+            if (s_specialFolders == null)
+            {
+                Interlocked.CompareExchange(ref s_specialFolders, new Dictionary<SpecialFolder, string>(), null);
+            }
+
+            string path;
+            lock (s_specialFolders)
+            {
+                if (!s_specialFolders.TryGetValue(folder, out path))
+                {
+                    path = GetSpecialFolder(folder);
+                    s_specialFolders[folder] = path;
+                }
+            }
```

Comment:
We don't need to lock for the happy case where we have the value cached, the worst that can happen in the other case is that we retrieve/store the same value twice if two threads are racing:

```suggestion
            if (!s_specialFolders.TryGetValue(folder, out path))
            {
                lock (s_specialFolders)
                {
                    path = GetSpecialFolder(folder);
                    s_specialFolders[folder] = path;
                }
            }
```

Reply:
@akoeplinger `TryGetValue` is not thread safe, if one tread is doing Add and another `TryGetValue`. It's not ConcurrentDictionary after all

Reply:
Hmm yeah, true. I'll need to stop reviewing threading code before my morning coffee :smile:

Reply:
@akoeplinger I am not a thread-safe expert neither though :)